crates/forge-runner/src/backtrace.rs

```diff
@@ -0,0 +1,226 @@
+use cairo_lang_starknet_classes::contract_class::ContractClass;
+use cheatnet::runtime_extensions::forge_runtime_extension::contracts_data::ContractsData;
+use cheatnet::state::EncounteredError;
+use indoc::indoc;
+use itertools::Itertools;
+use rayon::prelude::*;
+use starknet_api::core::ClassHash;
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::{env, fmt};
+
+const BACKTRACE_ENV: &str = "SNFORGE_BACKTRACE";
+
+pub fn add_backtrace_footer(
+    message: String,
+    contracts_data: &ContractsData,
+    encountered_errors: &[EncounteredError],
+) -> String {
+    if encountered_errors.is_empty() {
+        return message;
+    }
+
+    let is_backtrace_enabled = env::var(BACKTRACE_ENV).is_ok_and(|value| value == "1");
+    if !is_backtrace_enabled {
+        return format!(
+            "{message}\nnote: run with `{BACKTRACE_ENV}=1` environment variable to display a backtrace",
+        );
+    }
+
+    BacktraceContractRepository::new(contracts_data, encountered_errors)
+        .map(|repository| {
+            encountered_errors
+                .iter()
+                .filter_map(|error| repository.get_backtrace(error.pc, error.class_hash))
+                .map(|backtrace| backtrace.to_string())
+                .collect::<Vec<_>>()
+                .join("\n")
+        })
+        .map_or_else(
+            |err| format!("{message}\nfailed to create backtrace: {err}"),
+            |backtraces| format!("{message}\n{backtraces}"),
+        )
+}
+
+struct ContractBacktraceData {
+    contract_name: String,
+    casm_debug_info_start_offsets: Vec<usize>,
+    coverage_annotations: CoverageAnnotationsV1,
+    profiler_annotations: ProfilerAnnotationsV1,
+}
+
+impl ContractBacktraceData {
+    fn new(class_hash: &ClassHash, contracts_data: &ContractsData) -> Result<Self> {
+        let contract_name = contracts_data
+            .get_contract_name(class_hash)
+            .context(format!(
+                "failed to get contract name for class hash: {class_hash}"
+            ))?
+            .clone();
+
+        let contract_artifacts = contracts_data
+            .get_artifacts(&contract_name)
+            .context(format!(
+                "failed to get artifacts for contract name: {contract_name}"
+            ))?;
+
+        let contract_class = serde_json::from_str::<ContractClass>(&contract_artifacts.sierra)?;
+
+        let sierra_debug_info = contract_class
+            .sierra_program_debug_info
+            .as_ref()
+            .context("debug info not found")?;
+
+        let VersionedCoverageAnnotations::V1(coverage_annotations) =
+            VersionedCoverageAnnotations::try_from_debug_info(sierra_debug_info).context(indoc! {
+                "perhaps the contract was compiled without the following entry in Scarb.toml under [profile.dev.cairo]:
+                unstable-add-statements-code-locations-debug-info = true
+
+                or scarb version is less than 2.8.0
```

Comment:
A similar implementation has already been done for [coverage](https://github.com/foundry-rs/starknet-foundry/blob/5bbfa804a1d7aac97a9df9d35b1fde3f2c08cc6d/crates/forge-runner/src/coverage_api.rs#L70). 

However, to achieve this, we need metadata for the operation. I haven’t implemented this because it would either require:
- Propagating the metadata throughout the entire codebase up until this point, or
- Implementing it at the start of the `snforge` process, which is not optimal as you could run the backtrace feature without encountering any error. In such cases, the annotations wouldn’t even be needed, yet the process would still fail.

Reply:
Please create a task for this and we can resolve the details there.

Reply:
https://github.com/foundry-rs/starknet-foundry/issues/2710